### PR TITLE
Fix security bugs

### PR DESCRIPTION
**Pull request by Ishiihara**
@nehanarkhede @ewencp @gwenshap @miguno Sorry for the last minute change. But I have to get this in. Note: This PR only includes fix for secure HDFS, the issue connecting to secure Hive metastore is still under investigation.  

I performed a manual test of the HDFS connector on a multi-node cluster with a real KDC and it discovers some bugs in the security code. The Hadoop and Hive versions used in test are Hadoop 2.5.0 CDH 5.3.6 and Hive 0.13.1 CDH 5.3.6. 

The connector works smoothly with secure HDFS, however, the connector has issues with connecting to a secure Hive metastore. The following messages in the Hive metastore log when the connector tries to connect the secure Hive metastore: 

``
15/12/06 13:15:48 [pool-3-thread-21]: DEBUG security.UserGroupInformation: PrivilegedAction as:hive/worker4@LOCAL.DOMAIN (auth:KERBEROS) from:org.apache.hadoop.hive.thrift.HadoopThriftAuthBridge20S$Server$TUGIAssumingTransportFactory.getTransport(HadoopThriftAuthBridge20S.java:721)
15/12/06 13:15:48 [pool-3-thread-21]: DEBUG transport.TSaslServerTransport: transport map does not contain key
15/12/06 13:15:48 [pool-3-thread-21]: DEBUG transport.TSaslTransport: opening transport org.apache.thrift.transport.TSaslServerTransport@1f0ab9e1
15/12/06 13:15:48 [pool-3-thread-21]: DEBUG transport.TSaslTransport: SERVER: Writing message with status ERROR and payload length 19
15/12/06 13:15:48 [pool-3-thread-21]: DEBUG transport.TSaslServerTransport: failed to open server transport
org.apache.thrift.transport.TTransportException: Invalid status -128
	at org.apache.thrift.transport.TSaslTransport.sendAndThrowMessage(TSaslTransport.java:230)
	at org.apache.thrift.transport.TSaslTransport.receiveSaslMessage(TSaslTransport.java:184)
	at org.apache.thrift.transport.TSaslServerTransport.handleSaslStartMessage(TSaslServerTransport.java:125)
	at org.apache.thrift.transport.TSaslTransport.open(TSaslTransport.java:262)
	at org.apache.thrift.transport.TSaslServerTransport.open(TSaslServerTransport.java:41)
	at org.apache.thrift.transport.TSaslServerTransport$Factory.getTransport(TSaslServerTransport.java:216)
	at org.apache.hadoop.hive.thrift.HadoopThriftAuthBridge20S$Server$TUGIAssumingTransportFactory$1.run(HadoopThriftAuthBridge20S.java:724)
	at org.apache.hadoop.hive.thrift.HadoopThriftAuthBridge20S$Server$TUGIAssumingTransportFactory$1.run(HadoopThriftAuthBridge20S.java:721)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:356)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1608)
	at org.apache.hadoop.hive.thrift.HadoopThriftAuthBridge20S$Server$TUGIAssumingTransportFactory.getTransport(HadoopThriftAuthBridge20S.java:721)
	at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:227)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
15/12/06 13:15:48 [pool-3-thread-21]: ERROR server.TThreadPoolServer: Error occurred during processing of message.
java.lang.RuntimeException: org.apache.thrift.transport.TTransportException: Invalid status -128
	at org.apache.thrift.transport.TSaslServerTransport$Factory.getTransport(TSaslServerTransport.java:219)
	at org.apache.hadoop.hive.thrift.HadoopThriftAuthBridge20S$Server$TUGIAssumingTransportFactory$1.run(HadoopThriftAuthBridge20S.java:724)
	at org.apache.hadoop.hive.thrift.HadoopThriftAuthBridge20S$Server$TUGIAssumingTransportFactory$1.run(HadoopThriftAuthBridge20S.java:721)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:356)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1608)
	at org.apache.hadoop.hive.thrift.HadoopThriftAuthBridge20S$Server$TUGIAssumingTransportFactory.getTransport(HadoopThriftAuthBridge20S.java:721)
	at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:227)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.thrift.transport.TTransportException: Invalid status -128
	at org.apache.thrift.transport.TSaslTransport.sendAndThrowMessage(TSaslTransport.java:230)
	at org.apache.thrift.transport.TSaslTransport.receiveSaslMessage(TSaslTransport.java:184)
	at org.apache.thrift.transport.TSaslServerTransport.handleSaslStartMessage(TSaslServerTransport.java:125)
	at org.apache.thrift.transport.TSaslTransport.open(TSaslTransport.java:262)
	at org.apache.thrift.transport.TSaslServerTransport.open(TSaslServerTransport.java:41)
	at org.apache.thrift.transport.TSaslServerTransport$Factory.getTransport(TSaslServerTransport.java:216)
	... 10 more
`` 

I suspect that this is probably a Hive version mismatch issue as the Hive cli can access the secure metastore both locally and remotely with proper Kerberos credentials. I also tried to lower the version of Hive library to 0.13.1 in pom.xml but the same error still shows up.  I will investigate more and I really appreciate if anyone can help.

### CODE REVIEW

**Comment by miguno** [comment]
Do we need to add test(s) for this, @Ishiihara ?  There are none in the PR.

**Comment by Ishiihara** [comment]
@miguno It is hard to write unit tests for hadoop security. Many issues don't show up in single node cluster. I have to run manual tests on multi-node cluster.

**Comment by gwenshap** [comment]
@Ishiihara: You may want to open a JIRA to add a system test (which can have as many nodes as you need)

File: src/main/java/io/confluent/connect/hdfs/DataWriter.java [hunk 1]
```diff
@@ -109,19 +109,23 @@ public DataWriter(HdfsSinkConnectorConfig connectorConfig, SinkTaskContext conte
               + "the path to the keytab of the principal.");
         }
 
+        conf.set("hadoop.security.authentication", "kerberos");
+        conf.set("hadoop.security.authorization", "true");
```
Comment (gwenshap) [inline src/main/java/io/confluent/connect/hdfs/DataWriter.java]: Aren't those server side configurations?

File: src/main/java/io/confluent/connect/hdfs/DataWriter.java [hunk 2]
```diff
@@ -109,19 +109,23 @@ public DataWriter(HdfsSinkConnectorConfig connectorConfig, SinkTaskContext conte
...
         if (conf.get("dfs.namenode.kerberos.principal") == null) {
           conf.set("dfs.namenode.kerberos.principal", namenodePrincipal);
         }
+        log.info("Hadoop namenode principal: " + conf.get("dfs.namenode.kerberos.principal"));
 
-        final UserGroupInformation ugi = UserGroupInformation.loginUserFromKeytabAndReturnUGI(principal,
-                                                                                              keytab);
+        UserGroupInformation.setConfiguration(conf);
+        UserGroupInformation.loginUserFromKeytab(principal, keytab);
+        final UserGroupInformation ugi = UserGroupInformation.getLoginUser();
```
Comment (gwenshap) [inline src/main/java/io/confluent/connect/hdfs/DataWriter.java]: I'm a bit confused because you mentioned Hadoop 2.5.0, and I can't find these methods in the API (they exist in the older 1.0.x APIs).

It may be an Hadoop documentation issue? 
http://hadoop.apache.org/docs/r2.5.2/api/

File: src/main/java/io/confluent/connect/hdfs/hive/HiveMetaStore.java [hunk 3]
```diff
@@ -50,7 +50,9 @@ public HiveMetaStore(Configuration conf, HdfsSinkConnectorConfig connectorConfig
     HiveConf hiveConf = new HiveConf(conf, HiveConf.class);
     String hiveConfDir = connectorConfig.getString(HdfsSinkConnectorConfig.HIVE_CONF_DIR_CONFIG);
     String hiveMetaStoreURIs = connectorConfig.getString(HdfsSinkConnectorConfig.HIVE_METASTORE_URIS_CONFIG);
-    hiveConf.addResource(new Path(hiveConfDir, "hive-site.xml"));
+    if (!hiveConfDir.equals("")) {
+      hiveConf.addResource(new Path(hiveConfDir, "/hive-site.xml"));
+    }
```
Comment (gwenshap) [inline src/main/java/io/confluent/connect/hdfs/hive/HiveMetaStore.java]: Minor: Validate that hive-site.xml exists and print a warning if it doesn't? Looks like this can be a common mistake.

**Comment by gwenshap** [comment]
@Ishiihara: Does the patch actually solve the Hive issue? If not, what does it solve?

**Comment by Ishiihara** [comment]
@gwenshap  The connector uses kerberos for authentication even if the use didn’t provide the hadoop.conf.dir in the connector configuration. 2. The UserGroupInformation get the correct configuration by calling UserGroupInformation.setConfiguration(conf).

File: src/main/java/io/confluent/connect/hdfs/DataWriter.java (hunk 1 above)
Comment (Ishiihara) [inline src/main/java/io/confluent/connect/hdfs/DataWriter.java]: This makes sure that the connector uses kerberos for authentication even if the use didn’t provide the hadoop.conf.dir in the connector configuration

**Comment by Ishiihara** [comment]
@nehanarkhede @gwenshap @ewencp @miguno Gwen has verified the patch works with secure HDFS and Hive Metastore. The Hive version is hive 1.1.0 CDH 5.5.0.  Also modified the patch per Gwen's suggestions. Should be good to merge.

**Comment by gwenshap** [comment]
LGTM.

Validated on RHEL7 with CDH5.5.
